Enable non-safetensor ser/deser for TorchAoConfig quantized model 🔴

jerryzh168 opened the pull request:
Summary:
After https://github.com/huggingface/huggingface_hub/pull/2440 we added non-safetensor serialization and deserialization in huggingface, with this we can now add the support in transformers

Note that we don't plan to add safetensor serialization due to different goals of wrapper tensor subclass and safetensor see README for more details

Test Plan:
tested locally
https://gist.github.com/jerryzh168/965ccdbd595c9210d49cfbe31dc6705f

Reviewers:

Subscribers:

Tasks:

Tags:

jerryzh168 committed:
Enable non-safetensor serialization and deserialization for TorchAoConfig quantized model

Summary:
After https://github.com/huggingface/huggingface_hub/pull/2440 we added non-safetensor serialization and deserialization
in huggingface, with this we can now add the support in transformers

Note that we don't plan to add safetensor serialization due to different goals of wrapper tensor subclass and safetensor
see README for more details

Test Plan:
tested locally

Reviewers:

Subscribers:

Tasks:

Tags:

jerryzh168 committed:
formatting

jerryzh168 committed:
formatting

jerryzh168 committed:
minor fix

jerryzh168 committed:
formatting

jerryzh168 commented:
cc @SunMarc @Wauplin can you take a look?

SunMarc commented on docs/source/en/quantization/torchao.md:
```diff
@@ -40,6 +41,51 @@ quantized_model = torch.compile(quantized_model, mode="max-autotune")
 
 output = quantized_model.generate(**input_ids, max_new_tokens=10)
 print(tokenizer.decode(output[0], skip_special_tokens=True))
+
+# benchmark the performance
+import torch.utils.benchmark as benchmark
+
+def benchmark_fn(f, *args, **kwargs):
+    # Manual warmup
+    for _ in range(5):
+        f(*args, **kwargs)
+        
+    t0 = benchmark.Timer(
+        stmt="f(*args, **kwargs)",
+        globals={"args": args, "kwargs": kwargs, "f": f},
+        num_threads=torch.get_num_threads(),
+    )
+    return f"{(t0.blocked_autorange().mean):.3f}"
+
+MAX_NEW_TOKENS = 1000
+print("int4wo-128 model:", benchmark_fn(quantized_model.generate, **input_ids, max_new_tokens=MAX_NEW_TOKENS))
+
+bf16_model = AutoModelForCausalLM.from_pretrained(model_name, device_map="cuda", torch_dtype=torch.bfloat16)
+bf16_model = torch.compile(bf16_model, mode="max-autotune")
+print("bf16 model:", benchmark_fn(bf16_model.generate, **input_ids, max_new_tokens=MAX_NEW_TOKENS))
+
```
When torch 2.5 is out, we can add a script to show how to compile the forward with `mode="reduce-overhead"` and `fullgraph=True` + static cache. 

https://gist.github.com/SunMarc/2d863109027c083e5ea4d1c9b4991388

jerryzh168 replied:
should I attach this script to README now?

SunMarc replied:
Not on this PR ! We can do a follow up PR when torch 2.5 release day will be near

SunMarc commented on src/transformers/modeling_utils.py:
```diff
@@ -4430,7 +4440,15 @@ def _find_mismatched_keys(
                     continue
-                state_dict = load_state_dict(shard_file, is_quantized=is_quantized)
+                map_location = None
+                if (
+                    device_map is not None
+                    and "" in device_map
+                    and is_quantized
+                    and isinstance(hf_quantizer, TorchAoHfQuantizer)
+                ):
+                    map_location = torch.device(device_map[""])
```
WDYT ? This way, it will also work for multi-gpu case since device_map will be a dict + check that we are in `int4_weight_only` case 
```suggestion
                if (
                    device_map is not None
                    hf_quantizer is not None
                    and hf_quantizer.quantization_config.quant_method == QuantizationMethod.TORCHAO
                    and hf_quantizer.quant_type = "int4_weight_only"
                ):
                    map_location = torch.device([d for d in device_map.values() if d not in ["cpu", "disk"]][0])
```

SunMarc commented on src/transformers/modeling_utils.py:
```diff
@@ -56,6 +56,7 @@
     prune_linear_layer,
 )
 from .quantizers import AutoHfQuantizer, HfQuantizer
+from .quantizers.quantizer_torchao import TorchAoHfQuantizer
```
Not needed anymore if we check the QuantizationMethod instead 
```suggestion
```

SunMarc commented on src/transformers/quantizers/quantizer_torchao.py:
```diff
@@ -80,11 +80,11 @@ def update_torch_dtype(self, torch_dtype):
-                    "Setting torch_dtype to torch.bfloat16 for int4_weight_only quantization since only bfloat16 is supported right now. Please set torch_dtype=torch.bfloat16 to remove this warning."
-                )
-                torch_dtype = torch.bfloat16
+        if torch_dtype is None:
+            logger.warning_once(
+                "Setting torch_dtype to torch.bfloat16 for int4_weight_only quantization since only bfloat16 is supported right now. Please set torch_dtype=torch.bfloat16 to remove this warning."
+            )
+            torch_dtype = torch.bfloat16
```
Please leave the indent. This is only the case for int4_weight_only no ?

jerryzh168 replied:
I found the dtype is None at some point actually, so it feels like we'd need to set torch_dtype here?

SunMarc replied:
The dtype will be set to torch.float32 later on if the user don't specify `torch_dtype`. Let's leave the following code only for `int4_weight_only`

```py
        if torch_dtype is None:
            logger.warning_once(
                "Setting torch_dtype to torch.bfloat16 for int4_weight_only quantization since only bfloat16 is supported right now. Please set torch_dtype=torch.bfloat16 to remove this warning."
            )
            torch_dtype = torch.bfloat16
```

SunMarc commented on src/transformers/utils/quantization_config.py:
```diff
@@ -1082,6 +1081,23 @@ def get_loading_attributes(self):
         return loading_attibutes_dict
 
 
+def _get_torchao_quant_type_to_method():
+    if is_torchao_available():
+        from torchao.quantization import (
+            int4_weight_only,
+            int8_dynamic_activation_int8_weight,
+            int8_weight_only,
+        )
+
+        return {
+            "int4_weight_only": int4_weight_only,
+            "int8_weight_only": int8_weight_only,
+            "int8_dynamic_activation_int8_weight": int8_dynamic_activation_int8_weight,
+        }
+    else:
+        raise ValueError("TorchAoConfig requires torchao to be installed, please install with `pip install torchao`")
+
+
```
Can you put that in the `TorchAOConfig` class ? 
Otherwise, we can keep that we did before but we just need to make sure to pop `_STR_TO_METHOD` key when converting to a dict. The following should work: 


```py 
    def to_dict(self) -> Dict[str, Any]:
        """
        Serializes this instance to a Python dictionary. Returns:
            `Dict[str, Any]`: Dictionary of all the attributes that make up this configuration instance.
        """
        output = copy.deepcopy(self.__dict__)
        # not json serializable
        output.pop("_STR_TO_METHOD", None)

        return output
```

jerryzh168 replied:
OK will move to `TorchAOConfig` class

SunMarc commented on src/transformers/quantizers/quantizer_torchao.py:
```diff
@@ -162,7 +162,7 @@ def _process_model_after_weight_loading(self, model):
 
     @property
     def is_serializable(self):
-        return False
+        return True
```
Let's create a new arg for `is_serializable` called `safe_serialization` and we set it to `None` by default. This way, we return a warning if the user have set `safe_serialization=True ` for this quantizer. Note that in `save_pretrained`, an error will be raise afterwards: 

```py
        quantization_serializable = (
            hf_quantizer is not None and isinstance(hf_quantizer, HfQuantizer) and hf_quantizer.is_serializable
        )

        if hf_quantizer is not None and not _hf_peft_config_loaded and not quantization_serializable:
            raise ValueError(
                f"The model is quantized with {hf_quantizer.quantization_config.quant_method} and is not serializable - check out the warnings from"
                " the logger on the traceback to understand the reason why the quantized model is not serializable."
            )
```

Note that this will require to update this method for all quantization methods. LMK if this sounds good to you.

jerryzh168 replied:
yeah sounds good, what does `safe_serialization=None` mean btw?

SunMarc replied:
I mean just set the default value of this optional args to `None`.

jerryzh168 replied:
OK, I was trying to ask how we can handle None, right now I'm just treating it the same way we treat `False`, please let me know if we have to handle it differently

SunMarc commented on src/transformers/quantizers/quantizer_torchao.py:
```diff
@@ -162,7 +162,7 @@ def _process_model_after_weight_loading(self, model):
 
     @property
     def is_serializable(self):
-        return False
+        return True
```
Also we can add a check here to make sure that we have the required version of huggingface-hub to perform serialization.

jerryzh168 replied:
makes sense, I can use `0.25.0.dev0` (my local version), please let me know if there is a better version to check

SunMarc replied:
Looks like 0.25 is the version to check against ! The new version was release today: https://github.com/huggingface/huggingface_hub/releases/tag/v0.25.0

SunMarc reviewed (commented):
Thanks for your work @jerryzh168 to enable serialization ! Really appreciate that you are doing the PRs on huggingface hub and transformers ! This looks pretty good ! I left a few comments

jerryzh168 commented on docs/source/en/quantization/torchao.md:
```diff
@@ -40,6 +41,51 @@ quantized_model = torch.compile(quantized_model, mode="max-autotune")
+# save_to = "{user_id}/llama3-8b-int4wo-128"
+# quantized_model.push_to_hub(save_to, safe_serialization=False)
+
+# load quantized model
+ckpt_id = "llama3-8b-int4wo-128"  # or huggingface hub id
+loaded_quantized_model = AutoModelForCausalLM.from_pretrained(ckpt_id, device_map="cuda", use_safetensors=False, low_cpu_mem_usage=False)
```
sg, will do, yeah this is not up to date with the code I'm running I think

jerryzh168 committed:
address comments

jerryzh168 committed:
comments

jerryzh168 commented:
@SunMarc thanks for your thoughtful reviews! I have addressed all the comments I think, please take a look again, also not sure if the CI failure is relevant or not

jerryzh168 commented:
btw, current pytorch nightly has a perf regression: https://github.com/pytorch/ao/issues/898 and we hope to fix this before 2.5 cherry-pick deadline

SunMarc commented on src/transformers/modeling_utils.py:
```diff
@@ -2526,7 +2535,9 @@ def save_pretrained(
-            hf_quantizer is not None and isinstance(hf_quantizer, HfQuantizer) and hf_quantizer.is_serializable
+            hf_quantizer is not None
+            and isinstance(hf_quantizer, HfQuantizer)
+            and hf_quantizer.is_serializable(safe_serialization)
```
As this is an optional arg, I prefer to pass it like that.
```suggestion
            and hf_quantizer.is_serializable(safe_serialization=safe_serialization)
```

SunMarc reviewed (approved):
Thanks for iterating! LGTM! Just a nit. Let me know when the regression is fixed. I've pinged a core maintainer to review the PR

SunMarc commented:
To fix the failing test, can you rebase on main ?

jerryzh168 committed:
minor fix

jerryzh168 committed:
update doc

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/transformers/pr_33456). All of your documentation changes will be reflected on that endpoint. The docs are available until 30 days after the last update.

jerryzh168 commented:
@ArthurZucker can you take a look at the PR?

SunMarc commented:
I've merged recently PR adding a new quantizer. Sorry for that but could you rebase on main and update the `is_serializable` method ?

jerryzh168 committed:
refactor compressed tensor quantizer

jerryzh168 commented:
@SunMarc @ArthurZucker updated, please take a look again

ArthurZucker commented on src/transformers/quantizers/quantizer_torchao.py:
```diff
@@ -160,9 +160,19 @@ def _process_model_after_weight_loading(self, model):
-    @property
-    def is_serializable(self):
-        return False
+    def is_serializable(self, safe_serialization=None):
```
changing the property is a tad breaking, so let's just put the 🔴 on the PR!

ArthurZucker reviewed (approved):
Thanks for this PR, super sorry for the delay! 
Super important to have serialization!

ArthurZucker commented:
Thanks a lot @jerryzh168 🤗 great contributions and I love that we can upload serialized quantized weights to the hub now!
